File: TOTD/c/TOTD.c

#include "TOTD.h"

#include <stdio.h>
#include <stdlib.h>

double TOTD_dot(double * v_one, double * v_two, size_t size)
{
    double sum = 0.0;
    for (size_t i = 0; i < size; i++) sum += v_one[i] * v_two[i];
    return sum;
}

TOTD * TOTD_init(double * theta,
                 size_t size,
                 double * phi)
{
    // create new agent
    TOTD * agent;
    if ((agent = (TOTD *) malloc(sizeof(TOTD))) == NULL)
    {
        perror("malloc for new agent failed");
        return NULL;
    }

    // set passed in parameters
    agent -> size = size;
    agent -> theta = theta;

    // initialize V_old
    agent -> V_old = 0.0;

    // make phi
    if ((agent -> phi = (double *) malloc(size * sizeof(double))) == NULL)
    {
        perror("malloc for new agent phi failed");
        free(agent);
        return NULL;
    }
    for (size_t i = 0; i < size; i++) agent -> phi[i] = phi[i];

    // make e
    if ((agent -> e = (double *) calloc(size, sizeof(double))) == NULL)
    {
        perror("calloc for new agent e failed");
        free(agent -> phi);
        free(agent);
        return NULL;
    }

    return agent;
}

void TOTD_update(TOTD * agent,
                 double * phi_prime,
                 double reward,
                 double alpha,
                 double lambda,
                 double gamma)
{
    // calculate V and V_prime
    double V = TOTD_dot(agent -> phi, agent -> theta, agent -> size);
    double V_prime = TOTD_dot(phi_prime, agent -> theta, agent -> size);

    // calculate delta
    double delta = reward + gamma * V_prime - V;

    // update eligibility traces
    double e_phi = TOTD_dot(agent -> phi, agent -> e, agent -> size);
    for (size_t i = 0; i < agent -> size; i++)
    {
        agent -> e[i] = gamma * lambda * agent -> e[i] + agent -> phi[i] -
                         alpha * gamma * lambda * e_phi * agent -> phi[i];
    }

    // update theta
    for (size_t i = 0; i < agent -> size; i++)
    {
        agent -> theta[i] += alpha * (delta + V - agent -> V_old) * agent -> e[i] -
                             alpha * (V - agent -> V_old) * agent -> phi[i];
    }

    // update values
    agent -> V_old = V_prime;
    for (size_t i = 0; i < agent -> size; i++) agent -> phi[i] = phi_prime[i];
}

double TOTD_predict(TOTD * agent, double * phi)
{
    return TOTD_dot(phi, agent -> theta, agent -> size);
}

void TOTD_destroy(TOTD * agent)
{
    free(agent -> phi);
    free(agent -> e);
    free(agent);
}

File: TOTD/py/TOTD.py

import numpy as np

class TOTD:

    """
    Represents a true online temporal difference lambda learning agent.
    """

    def __init__(self, alpha, lambda_, gamma, theta, phi):
        """
        Constructs a new agent with the given parameters. Note that a copy of
        phi is created during the construction process.
        """
        self.alpha = np.array(alpha)
        self.old_alpha = np.zeros(self.alpha.shape)
        self.gamma = np.array(gamma)
        self.old_gamma = np.zeros(self.gamma.shape)
        self.lambda_ = np.array(lambda_)
        self.old_lambda = np.zeros(self.lambda_.shape)
        self.theta = np.atleast_2d(theta)
        self._phi = np.array(np.copy(phi)) # use a copy of phi
        self._e = np.zeros(np.shape(self.theta))
        self._V_old = 0

    def update(self, phi_prime, reward, alpha=None, lambda_=None, gamma=None):
        """
        Updates the parameter vector for a new observation. If any optional
        values are set then the new value of the optional is used for this and
        future calls that do not set the same optional value.
        """
        # set optional values
        if alpha is not None:
            self.alpha = np.array(alpha)
        if lambda_ is not None:
            self.lambda_ = np.array(lambda_)
        if gamma is not None:
            self.gamma = np.array(gamma)
        # calculate V and V_prime
        V = np.dot(self.theta, self._phi)
        V_prime = np.dot(self.theta, phi_prime)
        # calculate delta
        delta = reward + self.gamma * V_prime - V
        # update eligibility traces
        e_phi = np.dot(self._e, self._phi)
        self._e *= (self.old_lambda * self.old_gamma)[..., np.newaxis]
        self._e -= np.outer(self.old_lambda * self.old_gamma * self.old_alpha * e_phi, self._phi)
        self._e += self._phi
        # update theta
        self.theta += (self._e*(self.alpha*(delta + V - self._V_old))[..., np.newaxis])
        self.theta -= np.outer(self.alpha * (V - self._V_old), self._phi)
        # update values
        self._V_old = V_prime
        self._phi = np.array(np.copy(phi_prime))
        np.copyto(self.old_gamma, self.gamma)
        np.copyto(self.old_lambda, self.lambda_)
        np.copyto(self.old_alpha, self.alpha)

    def predict(self, phi):
        """
        Returns the current prediction for a given set of features phi.
        """
        return np.dot(self.theta, phi)
